sub/sd_sbr.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <assert.h>
+#include <math.h>
+#include <limits.h>
+
+#include <libavutil/common.h>
+#include <subrandr/subrandr.h>
+
+#include "mpv_talloc.h"
+
+#include "options/m_config.h"
+#include "options/options.h"
+#include "common/common.h"
+#include "demux/packet_pool.h"
+#include "demux/stheader.h"
+#include "sub/osd.h"
+#include "video/mp_image.h"
+#include "sd.h"
+
+struct sd_sbr_priv {
+    struct sbr_library *sbr_library;
+    struct sbr_renderer *sbr_renderer;
+    struct sbr_subtitles *sbr_subtitles;
+    struct mp_image_params video_params;
+    struct mp_osd_res osd;
+    struct sub_bitmaps *bitmaps;
+};
+
+static void enable_output(struct sd *sd, bool enable)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+    if (enable == !!ctx->sbr_renderer)
+        return;
+    if (ctx->sbr_renderer) {
+        sbr_renderer_destroy(ctx->sbr_renderer);
+        ctx->sbr_renderer = NULL;
+    } else {
+        ctx->sbr_renderer = sbr_renderer_create(ctx->sbr_library);
+    }
+}
+
+static int init(struct sd *sd)
+{
+    if(strcmp(sd->codec->codec, "textsub/srv3") && strcmp(sd->codec->codec, "textsub/vtt"))
```

Comment:
are we sure `sd->codec->codec` will always be valid at this point?

Reply:
No idea, should I just add an extra null check here?